src/packages/cli/src/detach.ts

```diff
@@ -0,0 +1,256 @@
+import { fork } from "child_process";
+import createInstanceName from "./process-name";
+import envPaths from "env-paths";
+import psList from "ps-list";
+import {
+  existsSync,
+  mkdirSync,
+  rmSync,
+  writeFileSync,
+  readdirSync,
+  readFileSync
+} from "fs";
+import path from "path";
+import { StartArgs } from "./types";
+import { FlavorName } from "@ganache/flavors";
+
+export type DetachedInstance = {
+  instanceName: string;
+  pid: number;
+  startTime: number;
+  host: string;
+  port: number;
+  flavor: FlavorName;
+  cmd: string;
+  version: string;
+};
+
+const dataPath = envPaths(`Ganache/instances`).data;
+if (!existsSync(dataPath)) mkdirSync(dataPath, { recursive: true });
+
+const READY_MESSAGE = "ready";
+
+const START_ERROR = "An error ocurred spawning a detached instance of Ganache:";
+
+/**
+ * Notify that the detached instance has started and is ready to receive requests.
+ */
+export function notifyDetachedInstanceReady() {
+  // in "detach" mode, the parent will wait until the "ready" message is
+  // received before disconnecting from the child process.
+  process.send(READY_MESSAGE);
+}
+
+/**
+ * Attempt to find and remove the instance file for a detached instance.
+ * @param  {number} pid the pid of the detached instance
+ * @returns boolean indicating whether the instance file was cleaned up successfully
+ */
+export function removeDetachedInstanceFile(pid: number): boolean {
+  const instanceFilename = `${dataPath}/${pid}`;
+  if (existsSync(instanceFilename)) {
+    rmSync(instanceFilename);
+    return true;
+  }
+  return false;
+}
+
+/**
+ * Attempts to stop a detached instance with the specified instance name by
+ * sending a SIGTERM signal. Returns a boolean indicating whether the process
+ * was found. If the PID is identified, but the process is not found, any
+ * corresponding instance file will be removed.
+ *
+ * Note: This does not guarantee that the instance actually stops.
+ * @param  {string} instanceName
+ * @returns boolean indicating whether the instance was found.
+ */
+export async function stopDetachedInstance(
+  instanceName: string
+): Promise<boolean> {
+  const instance = await findDetachedInstanceByName(instanceName);
+  if (instance !== undefined) {
+    try {
+      process.kill(instance.pid, "SIGTERM");
```

Comment:
Is it possible to `removeDetachedInstanceFile` here after we kill it rather than with the `process.on("exit"...` in the cli file? It would be nice to have the startup and shutdown/cleanup centralized in this one file.

Reply:
The reason that I do it here, is that we will clean up in the case that the process exits other than via `ganache instances stop`. However we also clean up any instance files when we attempt to access the list of instances.

Initially I preferred this, as we are cleaning up as early as possible, but after looking into it, I agree it's cleaner as you proposed.